### PR TITLE
add SaveOptimModel interface in analysis_predictor.h and test it in a…

### PR DESCRIPTION
**Pull request by lijianshe02**
在analysis_predictor.h中增加SaveOptimModel()接口，实现预测时模型优化后保存模型文件以及参数的功能，并在resnet50以及dam两个模型上增加了保存的优化后的模型与原始模型输出结果的对比测试

### CODE REVIEW

File: paddle/fluid/inference/api/analysis_predictor.cc [hunk 1]
```diff
@@ -770,6 +770,46 @@ std::string AnalysisPredictor::GetSerializedProgram() const {
   return inference_program_->Proto()->SerializeAsString();
 }
 
+// Add SaveOptimModel
+void AnalysisPredictor::SaveOptimModel(const std::string &dir) {
+  // save model
+  std::string model_name = dir + "/model";
+  std::ofstream outfile;
+  outfile.open(model_name.c_str(), std::ios::out | std::ios::binary);
```
Comment (fc500110) [inline paddle/fluid/inference/api/analysis_predictor.cc]: It's better to write like:
``` cpp
std::ofstream outfile(model_name, std::ios::out | std::ios::binary);
```

File: paddle/fluid/inference/api/analysis_predictor.cc [hunk 2]
```diff
@@ -770,6 +770,46 @@ std::string AnalysisPredictor::GetSerializedProgram() const {
...
+  std::string inference_prog_desc = GetSerializedProgram();
+  outfile << inference_prog_desc;
+  outfile.close();
```
Comment (fc500110) [inline paddle/fluid/inference/api/analysis_predictor.cc]: `fstream` is a RAII object. `fstream` will auto closing at the end of the scope and no need to call `close`.

File: paddle/fluid/inference/api/analysis_predictor.cc [hunk 3]
```diff
@@ -770,6 +770,46 @@ std::string AnalysisPredictor::GetSerializedProgram() const {
...
+  // save params
+  framework::ProgramDesc *save_program = new framework::ProgramDesc();
```
Comment (fc500110) [inline paddle/fluid/inference/api/analysis_predictor.cc]: Lack of `delete`, and there is no heap variable need to be used.

File: paddle/fluid/inference/api/analysis_predictor.cc [hunk 4]
```diff
@@ -770,6 +770,46 @@ std::string AnalysisPredictor::GetSerializedProgram() const {
...
+  framework::BlockDesc *save_block = save_program->MutableBlock(0);
+
+  const framework::ProgramDesc &main_program = program();
```
Comment (fc500110) [inline paddle/fluid/inference/api/analysis_predictor.cc:847]: use `auto` keyword properly to make code clean and easy to read.

**Review by fc500110 — CHANGES_REQUESTED** [review]

File: paddle/fluid/inference/api/analysis_predictor.cc (hunk 1 above)
Comment (lijianshe02) [reply paddle/fluid/inference/api/analysis_predictor.cc]: Done, thanks~

File: paddle/fluid/inference/api/analysis_predictor.cc (hunk 2 above)
Comment (lijianshe02) [reply paddle/fluid/inference/api/analysis_predictor.cc]: Done, thanks~

File: paddle/fluid/inference/api/analysis_predictor.cc (hunk 3 above)
Comment (lijianshe02) [reply paddle/fluid/inference/api/analysis_predictor.cc]: Done, thanks~

File: paddle/fluid/inference/api/analysis_predictor.cc (hunk 4 above)
Comment (lijianshe02) [reply paddle/fluid/inference/api/analysis_predictor.cc:847]: Done, thanks~

**Review by fc500110 — DISMISSED** [review]
LGTM

File: paddle/fluid/inference/tests/api/analyzer_dam_tester.cc [hunk 5]
```diff
@@ -313,5 +322,44 @@ TEST(Analyzer_dam, compare_determine) {
...
+                         const std::vector<std::vector<PaddleTensor>> &inputs) {
+  PrintConfig(orig_config, true);
+  PrintConfig(optim_config, true);
+  std::vector<PaddleTensor> orig_outputs, optim_outputs;
+  TestOneThreadPrediction(orig_config, inputs, &orig_outputs, false);
+  TestOneThreadPrediction(optim_config, inputs, &optim_outputs, false);
+  CompareResult(orig_outputs, optim_outputs);
+}
+
+TEST(Analyzer_resnet50, compare_optim_orig) {
```
Comment (wzzju) [inline paddle/fluid/inference/tests/api/analyzer_dam_tester.cc]: `Analyzer_resnet50` ? Do you want to name `Analyzer_dam`?

File: paddle/fluid/inference/tests/api/analyzer_dam_tester.cc [hunk 6]
```diff
@@ -313,5 +322,44 @@ TEST(Analyzer_dam, compare_determine) {
                        input_slots_all);
 }
 
+// Save optim model
+TEST(Analyzer_resnet50, save_optim_model) {
```
Comment (wzzju) [inline paddle/fluid/inference/tests/api/analyzer_dam_tester.cc]: `Analyzer_resnet50` ? Do you want to name `Analyzer_dam`?

File: paddle/fluid/inference/tests/api/analyzer_dam_tester.cc [hunk 7]
```diff
@@ -313,5 +322,44 @@ TEST(Analyzer_dam, compare_determine) {
...
+  AnalysisConfig cfg;
+  SetConfig(&cfg);
+  std::string optimModelPath =
+      FLAGS_infer_model.substr(0, FLAGS_infer_model.find_last_of("/")) +
+      "/saved_optim_model";
+  mkdir(optimModelPath.c_str(), 0777);
```
Comment (Superjomn) [inline paddle/fluid/inference/tests/api/analyzer_dam_tester.cc:334]: doc the magic number `0777`

File: paddle/fluid/inference/api/analysis_predictor.h [hunk 8]
```diff
@@ -86,6 +86,10 @@ class AnalysisPredictor : public PaddlePredictor {
 
   bool MkldnnQuantize();
 
+  // protobuf -> model
+  // params
```
Comment (luotao1) [inline paddle/fluid/inference/api/analysis_predictor.h]: 注释能否详细点？
protobuf->model, params是什么意思呢？是说program_desc存储成model？

File: paddle/fluid/inference/tests/api/analyzer_dam_tester.cc (hunk 5 above)
Comment (luotao1) [reply paddle/fluid/inference/tests/api/analyzer_dam_tester.cc]: 此处还没改

File: paddle/fluid/inference/tests/api/analyzer_dam_tester.cc (hunk 6 above)
Comment (lijianshe02) [reply paddle/fluid/inference/tests/api/analyzer_dam_tester.cc]: Done, thanks~

File: paddle/fluid/inference/tests/api/analyzer_dam_tester.cc (hunk 7 above)
Comment (lijianshe02) [reply paddle/fluid/inference/tests/api/analyzer_dam_tester.cc:334]: Done, thanks~

File: paddle/fluid/inference/api/analysis_predictor.h (hunk 8 above)
Comment (lijianshe02) [reply paddle/fluid/inference/api/analysis_predictor.h]: 是这个意思，添加了详细的注释了

**Review by fc500110 — APPROVED** [review]
LGTM

**Pull request merged**